mmengine/model/wrappers/test_time_aug.py

```diff
@@ -0,0 +1,163 @@
+# Copyright (c) OpenMMLab. All rights reserved.
+from abc import abstractmethod
+from typing import Dict, List, Optional, Union
+
+import torch
+
+from mmengine import MODELS
+from mmengine.optim import OptimWrapper
+from mmengine.structures import BaseDataElement
+from ..base_model import BaseDataPreprocessor, BaseModel
+
+# multi-batch inputs processed by different augmentations from the same batch.
+TestTimeAugInputs = List[Union[torch.Tensor, List[torch.Tensor]]]
+# multi-batch data samples processed by different augmentations from the same
+# batch. Each element in `EnhancedDataSamples` means an enhanced batch data.
+EnhancedDataSamples = List[List[BaseDataElement]]
+MergedDataSamples = List[BaseDataElement]
+
+
+@MODELS.register_module()
+class BaseTTAModel(BaseModel):
+    """Base model for inference with test-time augmentation.
+
+    ``BaseTTAModel`` is a wrapper for inference given multi-batch data.
+    It implements the :meth:`forward` for multi-batch data inference.
+    `multi-batch` data means data processed by different augmentation
+    from the same batch.
+
+    All subclasses should implement :meth:`merge_preds` for results fusion.
+
+    During test time augmentation, the data processed by
+    :obj:`mmcv.transforms.TestTimeAug`, and then collated by
+    `pseudo_collate` will have the following format:
+
+    .. code-block::
+
+        result = dict(
+            inputs=[
+                [image1_aug1, image2_aug1],
+                [image1_aug2, image2_aug2]
+            ],
+            data_samples=[
+                [data_sample1_aug1, data_sample2_aug1],
+                [data_sample1_aug2, data_sample2_aug2],
+            ]
+        )
+
+    ``image{1}_aug{1}`` means the 1st image of the batch, which is
+    augmented by the 1st augmentation.
+
+    ``BaseTTAModel`` will collate the data to:
+
+     .. code-block::
+
+        data1 = dict(
+            inputs=[image1_aug1, image2_aug1],
+            data_samples=[data_sample1_aug1, data_sample2_aug1]
+        )
+
+        data2 = dict(
+            inputs=[image1_aug2, image2_aug2],
+            data_samples=[data_sample1_aug2, data_sample2_aug2]
+        )
+
+    ``data1`` and ``data2`` will be passed to model, and the results will be
+    merged by :meth:`merge_preds`
+
+    Note:
+        :meth:`merge_results` is an abstract method, all subclasses should
+        implement it.
+
+    Args:
+        module (BaseModel): Tested model.
+        data_preprocessor (BaseDataPreprocessor or dict, optional): The
+            pre-process config For :class:`BaseDataPreprocessor`.
+    """
+
+    def __init__(
+            self,
+            module: BaseModel,
+            data_preprocessor: Optional[Union[dict,
+                                              BaseDataPreprocessor]] = None):
+
+        super().__init__(module)
```

Comment:
BaseModel does not accept `module` argument so should this `module` be changed to `data_preprocessor`?